src/cgutils.cpp

```diff
@@ -1623,34 +1639,52 @@ static Value *boxed(const jl_cgval_t &vinfo, jl_codectx_t *ctx, bool gcrooted)
         box = call_with_unsigned(box_char_func, as_value(t, vinfo));
     else if (jb == jl_ssavalue_type) {
         unsigned zero = 0;
-        v = as_value(t, vinfo);
+        Value *v = as_value(t, vinfo);
         assert(v->getType() == jl_ssavalue_type->struct_decl);
         v = builder.CreateExtractValue(v, makeArrayRef(&zero, 1));
         box = call_with_unsigned(box_ssavalue_func, v);
     }
-    else if (!jl_isbits(jt) || !jl_is_leaf_type(jt)) {
-        assert("Don't know how to box this type" && false);
-        return NULL;
-    }
     else if (!jb->abstract && jl_datatype_nbits(jb) == 0) {
+        // singleton
         assert(jb->instance != NULL);
         return literal_pointer_val(jb->instance);
     }
-    else {
+    return box;
+}
+
+// this is used to wrap values for generic contexts, where a
+// dynamically-typed value is required (e.g. argument to unknown function).
+// if it's already a pointer it's left alone.
+static Value *boxed(const jl_cgval_t &vinfo, jl_codectx_t *ctx, bool gcrooted)
+{
+    jl_value_t *jt = vinfo.typ;
+    if (jt == jl_bottom_type || jt == NULL)
+        // We have an undef value on a (hopefully) dead branch
+        return UndefValue::get(T_pjlvalue);
+    if (vinfo.constant)
+        return literal_pointer_val(vinfo.constant);
+    assert(vinfo.V);
+    if (vinfo.isboxed)
+        return vinfo.V;
+
+    assert(jl_isbits(jt) && jl_is_leaf_type(jt) && "This type shouldn't have been unboxed.");
+    Type *t = julia_type_to_llvm(jt);
+    assert(!type_is_ghost(t)); // should have been handled by isghost above!
```

Comment:
It's not clear to me which part of the code this comment about `isghost` refers to.